File: Sources/App/Classes/IRC/Users/IRCUserRelations.m
/* *********************************************************************
 *                  _____         _               _
 *                 |_   _|____  _| |_ _   _  __ _| |
 *                   | |/ _ \ \/ / __| | | |/ _` | |
 *                   | |  __/>  <| |_| |_| | (_| | |
 *                   |_|\___/_/\_\\__|\__,_|\__,_|_|
 *
 * Copyright (c) 2010 - 2018 Codeux Software, LLC & respective contributors.
 *       Please see Acknowledgements.pdf for additional information.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 *  * Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 *  * Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *  * Neither the name of Textual, "Codeux Software, LLC", nor the
 *    names of its contributors may be used to endorse or promote products
 *    derived from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 *********************************************************************** */

#import <os/lock.h>

#import "IRCChannel.h"
#import "IRCUserRelationsPrivate.h"

NS_ASSUME_NONNULL_BEGIN

@interface IRCUserRelations ()
@property(nonatomic, strong) NSMutableDictionary<IRCChannel *, IRCChannelUser *> *relationsPrivate;
@end

@implementation IRCUserRelations
{
	os_unfair_lock _relationsLock;
}

- (instancetype)init
{
	if ((self = [super init])) {
		self->_relationsLock = OS_UNFAIR_LOCK_INIT;

		self.relationsPrivate = [NSMutableDictionary dictionary];
	}

	return self;
}

- (NSDictionary<IRCChannel *, IRCChannelUser *> *)relations
{
	os_unfair_lock_lock(&self->_relationsLock);

	NSDictionary *relations = [self.relationsPrivate copy];

	os_unfair_lock_unlock(&self->_relationsLock);

	return relations;
}

- (NSArray<IRCChannel *> *)relatedChannels
{
	os_unfair_lock_lock(&self->_relationsLock);

	NSArray *relatedChannels = self.relationsPrivate.allKeys;

	os_unfair_lock_unlock(&self->_relationsLock);

	return relatedChannels;
}

- (NSArray<IRCChannelUser *> *)relatedUsers
{
	os_unfair_lock_lock(&self->_relationsLock);

	NSArray *relatedUsers = self.relationsPrivate.allValues;

	os_unfair_lock_unlock(&self->_relationsLock);

	return relatedUsers;
}

- (void)enumerateRelations:(void(NS_NOESCAPE ^)(IRCChannel *channel, IRCChannelUser *member, BOOL *stop))block
{
	/* Enumerate a snapshot so that the caller's block is free to
	 call back into this object without deadlocking. */
	NSDictionary *relations = self.relations;

	[relations enumerateKeysAndObjectsUsingBlock:block];
}

- (NSUInteger)numberOfRelations
{
	os_unfair_lock_lock(&self->_relationsLock);

	NSUInteger numberOfRelations = self.relationsPrivate.count;

	os_unfair_lock_unlock(&self->_relationsLock);

	return numberOfRelations;
}

- (void)associateUser:(IRCChannelUser *)user withChannel:(IRCChannel *)channel
{
	NSParameterAssert(user != nil);
	NSParameterAssert(channel != nil);

	if (channel.isChannel == NO) {
		return;
	}

	os_unfair_lock_lock(&self->_relationsLock);

	/* IRCChannel does not really support copying. It returns self.
	 The protocol is declared here in a cast, instead of in the
	 header for IRCChannel, so plugin author's don't make a mistake. */
	self.relationsPrivate[(IRCChannel<NSCopying> *)channel] = user;

	os_unfair_lock_unlock(&self->_relationsLock);
}

- (void)disassociateUserWithChannel:(IRCChannel *)channel
{
	NSParameterAssert(channel != nil);

	if (channel.isChannel == NO) {
		return;
	}

	os_unfair_lock_lock(&self->_relationsLock);

	[self.relationsPrivate removeObjectForKey:channel];

	os_unfair_lock_unlock(&self->_relationsLock);
}

- (nullable IRCChannelUser *)userAssociatedWithChannel:(IRCChannel *)channel
{
	NSParameterAssert(channel != nil);

	if (channel.isChannel == NO) {
		return nil;
	}

	os_unfair_lock_lock(&self->_relationsLock);

	IRCChannelUser *user = self.relationsPrivate[channel];

	os_unfair_lock_unlock(&self->_relationsLock);

	return user;
}

@end

NS_ASSUME_NONNULL_END

File: Sources/App/Modules/IRCProtocol/Presence/IRCUserRelations.swift
/* *********************************************************************
 *                  _____         _               _
 *                 |_   _|____  _| |_ _   _  __ _| |
 *                   | |/ _ \\ \/ / __| | | |/ _` | |
 *                   | |  __/>  <| |_| |_| | (_| | |
 *                   |_|\\___/_/\\_\\__|\\__,_|\\__,_|_|
 *
 * Copyright (c) 2010 - 2026 Codeux Software, LLC & respective contributors.
 *       Please see Acknowledgements.pdf for additional information.
 *
 *********************************************************************** */

import Foundation

@objc(IRCUserRelations)
public final class UserRelations: NSObject {
	private let lock = NSLock()
	private var storage: [IRCChannel: IRCChannelUser] = [:]

	@objc public var relations: [IRCChannel: IRCChannelUser] {
		withLock { storage }
	}

	@objc public var relatedChannels: [IRCChannel] {
		withLock { Array(storage.keys) }
	}

	@objc public var relatedUsers: [IRCChannelUser] {
		withLock { Array(storage.values) }
	}

	@objc public var numberOfRelations: UInt {
		withLock { UInt(storage.count) }
	}

	@objc(associateUser:withChannel:)
	public func associate(_ user: IRCChannelUser, with channel: IRCChannel) {
		guard channel.isChannel else {
			return
		}

		withLock {
			storage[channel] = user
		}
	}

	@objc(disassociateUserWithChannel:)
	public func disassociateUser(with channel: IRCChannel) {
		guard channel.isChannel else {
			return
		}

		withLock {
			storage.removeValue(forKey: channel)
		}
	}

	@objc(userAssociatedWithChannel:)
	public func userAssociated(with channel: IRCChannel) -> IRCChannelUser? {
		guard channel.isChannel else {
			return nil
		}

		return withLock { storage[channel] }
	}

	@objc(enumerateRelations:)
	public func enumerateRelations(
		_ block: (IRCChannel, IRCChannelUser, UnsafeMutablePointer<ObjCBool>) -> Void
	) {
		let snapshot = relations
		var stop = ObjCBool(false)

		withUnsafeMutablePointer(to: &stop) { stopPointer in
			for (channel, member) in snapshot {
				block(channel, member, stopPointer)

				if stopPointer.pointee.boolValue {
					break
				}
			}
		}
	}

	@discardableResult
	private func withLock<Result>(_ operation: () -> Result) -> Result {
		lock.lock()
		defer { lock.unlock() }

		return operation()
	}
}
